File: correct-crypto/ruby/signature.rb
# frozen_string_literal: true
# Correct Ed25519 signing and verification (OpenSSL 3).
require 'openssl'

def sign_and_verify(message)
  key = OpenSSL::PKey.generate_key('ED25519')
  signature = key.sign(nil, message)    # Ed25519 uses no separate digest
  key.verify(nil, signature, message)
end

File: correct-crypto/go/signature.go
package correctcrypto

import (
	"crypto/ed25519"
	"crypto/rand"
)

// SignAndVerify demonstrates correct Ed25519 signing and verification.
func SignAndVerify(message []byte) (bool, error) {
	public, private, err := ed25519.GenerateKey(rand.Reader)
	if err != nil {
		return false, err
	}
	signature := ed25519.Sign(private, message)
	return ed25519.Verify(public, message, signature), nil
}
